File: backend/fastapi-server/services/health_service.py
"""
Health service for checking the status of components
"""

import asyncio
import logging
from typing import Dict, Any

from utils.text_utils import sanitize_error_message
from models.schema import HealthStatus

# Configure logging
logger = logging.getLogger(__name__)


class HealthService:
    """Handles health check functionality"""
    
    def __init__(self, config: Dict[str, Any], chroma_client, llm_client):
        self.config = config
        self.chroma_client = chroma_client
        self.llm_client = llm_client
        self._last_status = None
        self._last_check_time = 0
        self._cache_ttl = 30  # Cache health status for 30 seconds
    
    async def get_health_status(self, use_cache: bool = True) -> HealthStatus:
        """Get health status of all components"""
        current_time = asyncio.get_event_loop().time()
        
        # Return cached status if available and not expired
        if use_cache and self._last_status and (current_time - self._last_check_time) < self._cache_ttl:
            return self._last_status
        
        status = {
            "status": "ok",
            "components": {
                "server": {
                    "status": "ok"
                },
                "chroma": {
                    "status": "unknown"
                },
                "llm": {
                    "status": "unknown"
                }
            }
        }
        
        # Check Chroma
        try:
            # Simple heartbeat check in thread pool
            await asyncio.get_event_loop().run_in_executor(
                None, 
                self.chroma_client.heartbeat
            )
            status["components"]["chroma"]["status"] = "ok"
        except Exception as e:
            status["components"]["chroma"]["status"] = "error"
            status["components"]["chroma"]["error"] = sanitize_error_message(str(e))
        
        # Check LLM (Ollama)
        try:
            llm_ok = await self.llm_client.verify_connection()
            status["components"]["llm"]["status"] = "ok" if llm_ok else "error"
            if not llm_ok:
                status["components"]["llm"]["error"] = "Failed to connect to Ollama"
        except Exception as e:
            status["components"]["llm"]["status"] = "error"
            status["components"]["llm"]["error"] = sanitize_error_message(str(e))
        
        # Overall status
        if any(component["status"] != "ok" for component in status["components"].values()):
            status["status"] = "error"
        
        # Cache the result
        self._last_status = HealthStatus(**status)
        self._last_check_time = current_time
        
        return self._last_status
    
    def is_healthy(self, health: HealthStatus) -> bool:
        """Check if the system is healthy based on health status"""
        return health.status == "ok"

File: backend/express/src/services/healthService.ts
import { ChromaClient } from 'chromadb';
import { AppConfig } from '../types';
import { OllamaClient } from '../clients/ollamaClient';
import { VLLMClient } from '../clients/vllmClient';
import { AudioService } from './audioService';

export interface HealthStatus {
  status: string;
  components: {
    [key: string]: {
      status: string;
      error?: string;
    };
  };
}

/**
 * Service for handling health checks
 */
export class HealthService {
  private config: AppConfig;
  private chromaClient: ChromaClient;
  private llmClient: OllamaClient | VLLMClient;
  private audioService: AudioService;
  private lastStatus: HealthStatus | null = null;
  private lastCheckTime = 0;
  private readonly cacheTtl = 30; // Cache health status for 30 seconds
  
  constructor(
    config: AppConfig, 
    chromaClient: ChromaClient, 
    llmClient: OllamaClient | VLLMClient,
    audioService: AudioService
  ) {
    this.config = config;
    this.chromaClient = chromaClient;
    this.llmClient = llmClient;
    this.audioService = audioService;
  }
  
  /**
   * Get health status of all components
   */
  async getHealthStatus(useCache: boolean = true): Promise<HealthStatus> {
    const currentTime = Date.now() / 1000;

    // Return cached status if available and not expired
    if (useCache && this.lastStatus && (currentTime - this.lastCheckTime) < this.cacheTtl) {
      return this.lastStatus;
    }

    const status: HealthStatus = {
      status: 'ok',
      components: {
        server: {
          status: 'ok'
        },
        chroma: {
          status: 'unknown'
        },
        llm: {
          status: 'unknown'
        }
      }
    };

    // Check Chroma
    try {
      await this.chromaClient.heartbeat();
      status.components.chroma.status = 'ok';
    } catch (error: any) {
      status.components.chroma.status = 'error';
      status.components.chroma.error = error.message;
    }

    // Check LLM (Ollama/vLLM)
    try {
      const llmOk = await this.llmClient.verifyConnection();
      status.components.llm.status = llmOk ? 'ok' : 'error';
      if (!llmOk) {
        status.components.llm.error = 'Failed to connect to LLM service';
      }
    } catch (error: any) {
      status.components.llm.status = 'error';
      status.components.llm.error = error.message;
    }

    // Overall status
    if (Object.values(status.components).some(component => component.status !== 'ok')) {
      status.status = 'error';
    }

    // Cache the result
    this.lastStatus = status;
    this.lastCheckTime = currentTime;

    return status;
  }
  
  /**
   * Determine if all services are healthy
   */
  isHealthy(health: HealthStatus): boolean {
    return health.status === 'ok';
  }
}